typeid/typeid-go/typeid.go

```diff
@@ -6,42 +6,58 @@ import (
 )
 
 // TypeID is a unique identifier with a given type as defined by the TypeID spec
-type TypeID[P PrefixType] struct {
-	prefix string
-	suffix string
+type TypeID struct {
+	value     string // The complete "prefix_suffix" string
+	prefixLen uint8  // Length of prefix for extracting parts
 }
 
 // Prefix returns the type prefix of the TypeID
-func (tid TypeID[P]) Prefix() string {
-	if isAnyPrefix[P]() {
-		return tid.prefix
+func (tid TypeID) Prefix() string {
+	if tid.prefixLen == 0 {
+		return ""
 	}
-	return defaultPrefix[P]()
+	return tid.value[:tid.prefixLen]
 }
 
-const zeroSuffix = "00000000000000000000000000"
+const ZeroSuffix = "00000000000000000000000000"
+
+// zeroID is a zero-value TypeID with empty prefix and zero UUID.
+var zeroID = TypeID{}
 
 // Suffix returns the suffix of the TypeID in it's canonical base32 representation.
-func (tid TypeID[P]) Suffix() string {
-	// We want to treat the "empty" TypeID as equivalent to the 'zero' typeid
-	if tid.suffix == "" {
-		return zeroSuffix
+func (tid TypeID) Suffix() string {
+	if tid.prefixLen == 0 {
+		// No prefix, entire value is the suffix
+		if tid.value == "" {
+			return ZeroSuffix
+		}
+		return tid.value
+	}
+	// Has prefix, suffix starts after prefix + underscore
+	if len(tid.value) <= int(tid.prefixLen)+1 {
+		return ZeroSuffix
+	}
+	suffix := tid.value[tid.prefixLen+1:]
+	if suffix == "" {
+		return ZeroSuffix
 	}
-	return tid.suffix
+	return suffix
 }
 
 // String returns the TypeID in it's canonical string representation of the form:
 // <prefix>_<suffix> where <suffix> is the canonical base32 representation of the UUID
-func (tid TypeID[P]) String() string {
-	if tid.Prefix() == "" {
-		return tid.Suffix()
+func (tid TypeID) String() string {
+	if tid.value == "" {
+		return ZeroSuffix
```

Comment:
Again, maybe we should adopt the semantics where a zero-value TypeID is invalid and distinct from the ZeroSuffix ID. https://go.dev/play/p/mahjT66CfiM